src/common.jl

```diff
@@ -298,6 +300,44 @@ function Base.iszero(p::AbstractPolynomial)
     return all(iszero.(coeffs(p))) && p[0] == 0
 end
 
+# See discussions in https://github.com/JuliaMath/Polynomials.jl/issues/258
+is_(fn, p::AbstractPolynomial) = all(fn, coeffs(p))  # Do not export
+as_(fn, p::P) where {P<:AbstractPolynomial} = ⟒(P)(fn.(coeffs(p)), p.var)  # Do not export
+
+"""
+    isreal(p::AbstractPolynomial)
+
+Determine whether a polynomial is a real polynomial, i.e., having only real numbers as coefficients.
+
+See also: [`real`](@ref)
+"""
+Base.isreal(p::AbstractPolynomial) = is_(isreal, p)
+"""
+    real(p::AbstractPolynomial)
+
+Construct a real polynomial from the real parts of the coefficients of `p`.
+
+See also: [`isreal`](@ref)
+
+!!! note
+    This could cause losing terms in `p`. This method is usually called on polynomials like `p = Polynomial([1, 2 + 0im, 3.0, 4.0 + 0.0im])` where you want to chop the imaginary parts of the coefficients of `p`.
+"""
+Base.real(p::AbstractPolynomial) = as_(real, p)
+
+"""
+    isintegral(p::AbstractPolynomial)
+
+Determine whether a polynomial is an integer polynomial, i.e., having only integers as coefficients.
+"""
+isintegral(p::AbstractPolynomial) = is_(isinteger, p)
+
+"""
+    ismonic(p::AbstractPolynomial)
+
+Determine whether a polynomial is a monic polynomial, i.e., its leading coefficient is one.
+"""
+ismonic(p::AbstractPolynomial) = isone(coeffs(p)[end])
```

Comment:
I add one function `ismonic` to demonstrate `isXXX` can mean a character that applies only to 1 coefficient, not all. So `is_` might not be a good name that covers all usage?

Reply:
Probably not. The basic pattern was `all(fn, coeffs(p))` so something with "all" seems more apt. In looking at the docs for `all` I find: `all(p, itr) -> Bool`. Perhaps it makes sense here to just define:

```
Base.all(f, p::AbstractPolynomial) = all(f, p[:])
```

Though then the idiomatic means for this would just be: `all(isinteger, p[:])`. With that realization, it doesn't really make much sense to introduce an `is_` fun at all.

(Unfortunately for this purpose, we have the convention that iteration yields the terms not the coefficients, so `p[:]` is needed, not just a `p`.)

So maybe, for pointing people in a generalizable direction we would define things like

```
Base.map(fn, p::P) where {P <: AbstractPolynomial} = ⟒(P)(map(fn, coeffs(p)), p.var)
isreal(p::P) where {P <: AbstractPolynomial} = all(isreal, p[:])
real(p::P) where {P <: AbstractPolynomial}  = map(real, p)
```

and if you really think exposing `isintegral` is needed, then

```
isintegral(p::P) = all(isinteger, p[:])
```

Though, I'd argue the `all` usage and the `isreal` example would suffice and `isintegral` could be documented in `isreal` as an example of the general pattern.

Thanks for playing around with this. I think it highlighted some things for me.

Finally, `ismonic` would be just `isone(p[end])`. (The `coeffs` call can be costly for SparsePolynomials, so best to avoid it when possible.) (Maybe `isone(p[degree[end]])` is more clear?)

Reply:
One concern rises to me that both `all` & `map` mentioned above tried to regard only the coefficients as the important part of a `Polynomial`. And then I found that `Polynomial`s with different `var`s cannot multiply. In that sense, should they be named `UnivariatePolynomial` and this package `UnivariatePolynomials`?

Reply:
I’d say that would have been a good observation way back when, but this was one of the very first packages for Julia.  So it isn’t going to change, just evolve it’s internals as people want new functionality.

Reply:
> Finally, ismonic would be just isone(p[end]). (The coeffs call can be costly for SparsePolynomials, so best to avoid it when possible.) (Maybe isone(p[degree[end]]) is more clear?)

Somehow I found `isone(coeffs(p)[end])` is much faster than `isone(p[degree(p)])` for regular `Polynomial`s. Am I missing something?

```julia
ulia> x = Polynomial([1, 2, 3, 4])
Polynomial(1 + 2*x + 3*x^2 + 4*x^3)

julia> using BenchmarkTools

julia> @btime ismonic(x)  # isone(coeffs(p)[end])
  15.899 ns (0 allocations: 0 bytes)
false

julia> @btime ismonic($x)  # isone(coeffs(p)[end])
  1.694 ns (0 allocations: 0 bytes)
false

julia> @btime ismonic(x)  # isone(p[degree(p)])
  68.000 ns (2 allocations: 128 bytes)
false

julia> @btime ismonic($x)  # isone(p[degree(p)])
  51.573 ns (2 allocations: 128 bytes)
false
```

Why is that?

Reply:
Except for SparsePolynomials, coeffs(p) is just p.coeffs, so adding degree will be slower, as degree has  check to return -1 if the polynomial is empty. (your ismonic will have to check that too). However, for sparsepolynomials, `coeffs` takes a sparse dictionary and generates the full vector of indices so could be very costly. I realize now, just using `p[end]` should work for this term, which avoids the branch in `degree`.

Reply:
Looks like just this last change and this will be a go. Good point made about `real(p)` dropping degree as a possibility.

Reply:
Updated. `p[end]` is much faster than `p[degree(p)]`, though still a little bit slower than `coeffs(p)[end]` on `Polynomial([1, 2, 3, 4])`.